samza-core/src/main/java/org/apache/samza/zk/ZkLock.java

```diff
@@ -0,0 +1,102 @@
+ * under the License.
+ */
+package org.apache.samza.zk;
+
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import org.apache.samza.SamzaException;
+import org.apache.samza.coordinator.DistributedLock;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Distributed lock primitive for Zookeeper.
+ */
+public class ZkLock implements DistributedLock {
+
+  public static final Logger LOG = LoggerFactory.getLogger(ZkLock.class);
+  private final ZkUtils zkUtils;
+  private final String lockPath;
+  private final String participantId;
+  private final ZkKeyBuilder keyBuilder;
+  private final Random random = new Random();
+  private String nodePath = null;
+
+  public ZkLock(String participantId, ZkUtils zkUtils, String initLockPath) {
+    this.zkUtils = zkUtils;
+    this.participantId = participantId;
+    this.keyBuilder = this.zkUtils.getKeyBuilder();
+    lockPath = String.format("%s/%s", keyBuilder.getRootPath(), initLockPath);
+    zkUtils.makeSurePersistentPathsExists(new String[] {lockPath});
+  }
+
+  /**
+   * Tries to acquire a lock in order to create intermediate streams. On failure to acquire lock, it keeps trying until the lock times out.
+   * Creates a sequential ephemeral node to acquire the lock. If the path of this node has the lowest sequence number, the processor has acquired the lock.
+   * @param timeout Duration of lock acquiring timeout.
+   * @param unit Unit of the timeout defined above.
+   * @return true if lock is acquired successfully, false if it times out.
+   */
+  @Override
+  public boolean lock(long timeout, TimeUnit unit) {
+    nodePath = zkUtils.getZkClient().createEphemeralSequential(lockPath + "/", participantId);
+
+    //Start timer for timeout
+    long startTime = System.currentTimeMillis();
+    long lockTimeout = TimeUnit.MILLISECONDS.convert(timeout, unit);
+
+    while ((System.currentTimeMillis() - startTime) < lockTimeout) {
+      List<String> children = zkUtils.getZkClient().getChildren(lockPath);
+      int index = children.indexOf(ZkKeyBuilder.parseIdFromPath(nodePath));
+
+      if (children.size() == 0 || index == -1) {
+        throw new SamzaException("Looks like we are no longer connected to Zk. Need to reconnect!");
+      }
+      // Acquires lock when the node has the lowest sequence number and returns.
+      if (index == 0) {
+        LOG.info("Acquired lock for participant id: {}", participantId);
+        return true;
+      } else {
+        try {
+          Thread.sleep(random.nextInt(1000));
+        } catch (InterruptedException e) {
+          Thread.interrupted();
+        }
+        LOG.info("Trying to acquire lock again...");
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Unlocks, by deleting the ephemeral sequential node created to acquire the lock.
+   */
+  @Override
+  public void unlock() {
+    if (nodePath != null) {
+      zkUtils.getZkClient().delete(nodePath);
```

Comment:
So, I would recommend to think it through about the zkClient lifecycle needed in this lock implementation. Are we assuming that the zkClient is initialized by some other class and just passed in to this lock implementation? Or are we assuming that the lock would internally own the lifecycle of the zkClient? If we are assuming the first case, I think that this DistributedLock implementation lack of handling on the failure cases: what if the zkClient connection to the ZK server was dropped *after* the user has called lock() and proceeded into the critical section? If the disconnect time is longer than the session timeout, there would be race conditions. How do the users of the DistributedLock being notified of this issue after calling the lock()?